Made URL of new channel creation page more generic

shruti862 opened the pull request:
## Summary
Made URL of new channel creation page more generic to avoid 404 channel not found message on refreshing the page
[Screencast from 22-12-24 12:22:47 AM IST.webm](https://github.com/user-attachments/assets/4f8fb64f-efed-4028-b9aa-24326ae5084b)

## References
Fixes issue #2389 
## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

…

bjester commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelModal.vue:
```diff
@@ -269,46 +269,81 @@
+            .then(channel => {
+              if (this.isNew) {
+                const newChannelId = channel.id;
+                window.location = window.Urls.channel(newChannelId);
```
I realize this line of code was existing, but ideally it should have used `window.location.assign`. Since the behavior of the new channel creation page is changing, I think it might be better to either use `window.location.replace` directly, or perhaps `replace` the current location with the `RouteNames.CHANNEL_EDIT` route using the new channel ID, then `assign` to `window.Urls.channel(newChannelId)`

bjester commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelModal.vue:
```diff
@@ -269,46 +269,81 @@
+              if (this.isNew) {
+                const newChannelId = channel.id;
+                window.location = window.Urls.channel(newChannelId);
+                this.$store.dispatch('showSnackbarSimple', this.$tr('channelCreated'));
```
Since the page needs a reload, I don't know that this snackbar is worth it. The display of it must race the page load, which could mean it's never shown, depending on various factors related to speed.

bjester commented on contentcuration/contentcuration/frontend/shared/vuex/channel/actions.js:
```diff
@@ -82,10 +82,12 @@ export function commitChannel(
+  if (id && context.state.channelsMap[id]) {
     if (!id) {
-      throw ReferenceError('id must be defined to update a channel');
+      throw new ReferenceError('id must be defined to update a channel');
```
nice catch

bjester commented on contentcuration/contentcuration/frontend/shared/vuex/channel/actions.js:
```diff
@@ -118,10 +120,48 @@ export function commitChannel(
+    }
+
+    // Log the data before sending it to create a new channel
+    console.log('Creating new channel with data:', newChannelData);
```
Please remove the `console.log` statements in this vuex action. Thank you

bjester commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelModal.vue:
```diff
@@ -269,46 +269,81 @@
+
+      // If there's no channelId (i.e., it's a new channel), skip verifications
+      if (!channelId) {
+        //this.isNew = true;
```
Please remove this commented out code.

bjester commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelModal.vue:
```diff
@@ -269,46 +269,81 @@
+      // If there's no channelId (i.e., it's a new channel), skip verifications
+      if (!channelId) {
+        //this.isNew = true;
+        this.header = 'New Channel'; // Default header for a new channel
```
This seems redundant with the code in `mounted`. I suggest removing this.

bjester commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelModal.vue:
```diff
@@ -269,46 +269,81 @@
-      this.header = this.channel.name; // Get channel name when user enters modal
+      if (this.isNew) {
+        // For a new channel, set a default header
+        this.header = 'New Channel';
```
This needs to use our UI strings. For example, using `this.$tr` with the preexisting string key in the `$trs` object of the component.

bjester commented on contentcuration/contentcuration/frontend/shared/vuex/channel/actions.js:
```diff
@@ -118,10 +120,48 @@ export function commitChannel(
+    return Channel.createModel(newChannelData).then(response => {
+      console.log('API Response:', response);
+
+      const createdChannel = response.data || response; // Adjust based on API structure
```
It must be one or the other-- either `response.data` or `response`? I suggest making this specific.

bjester commented on contentcuration/contentcuration/frontend/shared/vuex/channel/actions.js:
```diff
@@ -118,10 +120,48 @@ export function commitChannel(
+    }
+    if (contentDefaults !== NOVALUE) {
+      newChannelData.content_defaults = contentDefaults;
+    }
```
The above code is very similar to the `if` branch above in this function. I'm requesting deduplication of the code because the only real difference should be the presence of `id` which isn't significant enough to warrant the duplication. For example, it seems there are differences here in how it handles the thumbnail above, which leads to a regression that I can see when you upload a thumbnail for a new channel and adjust it's cropping.

bjester reviewed (changes requested):
Thanks for your contribution! I tested it and it works, but I have several requested changes and suggestions

shruti862 commented:
hey @bjester ,I’ve implemented the changes you suggested, and everything is working as expected in terms of the channel creation process. However, I’ve encountered another issue. Whenever I close the ChannelModal without creating a new channel or when selecting the "EXIT WITHOUT SAVING" option after editing, an empty channel is still being created. 
![Screenshot from 2025-01-23 16-07-18](https://github.com/user-attachments/assets/b0d3e30e-2f76-4300-aab6-3a7a9f2b3357)

I don't know why this is happening ,Could you kindly assist in identifying the root cause of this behavior and help me resolve it?

MisRob commented:
> However, I’ve encountered another issue. Whenever I close the ChannelModal without creating a new channel or when selecting the "EXIT WITHOUT SAVING" option after editing, an empty channel is still being created which on reloading the page vanishes.

@shruti862 was this the case before your latest changes in this pull request? If so, feel free to open a new issue if there is not one yet. If your changes introduce it, it'd be best to spend some time debugging and resolve it in this pull request.

shruti862 commented:
@MisRob, the empty channel creation issue is due to my recent changes. After debugging, I have identified the root cause and will resolve it soon.

shruti862 commented:
Hey, @MisRob @bjester , createChannel function was creating an empty channel ::
```
  newChannel() {
        this.$analytics.trackClick('channel_list', 'Create channel');
        this.createChannel().then(id => {
          this.$router.push({
            name: RouteNames.CHANNEL_EDIT,
            params: { channelId: id, tab: 'edit' },
            query: { last: this.$route.name },
          });
        });
      },
```
When I removed the function from the line of code:
```
newChannel() {
        this.$analytics.trackClick('channel_list', 'Create channel');
        
          this.$router.push({
            name: RouteNames.NEW_CHANNEL,
            query: { last: this.$route.name },
          });
      },
```
Due to this change ,one frontend test is failing which says new channel should be created when new channel button is clicked which should not be the case because clicking on create button in new channel modal should create a new channel.
![Screenshot from 2025-01-29 16-58-48](https://github.com/user-attachments/assets/7a1d2a86-703e-49cd-9783-348a1475e7cf)

shruti862 commented:
Hey @MisRob @bjester,

Could you assist me with the failing frontend test issue? I’d like to understand the cause and how to resolve it. Your guidance would be greatly appreciated.

Thanks!

MisRob commented:
Hello @shruti862, we will follow-up with you, however please be patient as we have a large queue of pull request to work through. It's good to send a reminder if we don't reply roughly after a week. Thank you!

shruti862 commented:
@MisRob ,I apologize if I caused any inconvenience. I appreciate your time and will wait patiently. Thank you!

MisRob commented:
No problem @shruti862 at all. Thank you.

MisRob commented:
Hi @shruti862

> the empty channel creation issue is due to my recent changes. After debugging, I have identified the root cause and will resolve it soon.

Good news, thanks.

> Due to this change ,one frontend test is failing which says new channel should be created when new channel button is clicked which should not be the case because clicking on create button in new channel modal should create a new channel

This sounds that you just need to update tests to align them well with your changes, is that right? Basically the test suite would ideally correspond to user experience so you're welcome to update the test accordingly.

MisRob commented:
@shruti862 Also if you find any other tests related to this feature that are perhaps still passing but their description or implementation doesn't correspond anymore, that's always great to update too. Thanks for reaching out to clarify this and let us know if you needed anything else.

shruti862 commented:
Hey @MisRob @bjester, I’ve implemented all the suggested changes and also fixed the failing frontend test. Whenever you have a moment, could you please take a look at my changes and let me know if any further improvements are needed? :)

bjester commented:
Thank you @shruti862! The changes look good and function appropriately

MisRob commented:
Thanks!
